climada/util/calibrate/bayesian_optimizer.py

```diff
@@ -0,0 +1,160 @@
+
+    This optimizer reports the target function value for each parameter set and
+    *maximizes* that value. Therefore, a higher target function value is better.
+    The cost function, however, is still minimized: The target function is defined as
+    the inverse of the cost function.
+
+    For details on the underlying optimizer, see
+    https://github.com/bayesian-optimization/BayesianOptimization.
+
+    Parameters
+    ----------
+    input : Input
+        The input data for this optimizer. See the Notes below for input requirements.
+    verbose : int, optional
+        Verbosity of the optimizer output. Defaults to 1.
+    random_state : int, optional
+        Seed for initializing the random number generator. Defaults to 1.
+    allow_duplicate_points : bool, optional
+        Allow the optimizer to sample the same points in parameter space multiple times.
+        This may happen if the parameter space is tightly bound or constrained. Defaults
+        to ``True``.
+    bayes_opt_kwds : dict
+        Additional keyword arguments passed to the ``BayesianOptimization`` constructor.
+
+    Notes
+    -----
+    The following requirements apply to the parameters of :py:class:`Input` when using
+    this class:
+
+    bounds
+        Setting ``bounds`` in the ``Input`` is required because the optimizer first
+        "explores" the bound parameter space and then narrows its search to regions
+        where the cost function is low.
+    constraints
+        Must be an instance of ``scipy.minimize.LinearConstraint`` or
+        ``scipy.minimize.NonlinearConstraint``. See
+        https://github.com/bayesian-optimization/BayesianOptimization/blob/master/examples/constraints.ipynb
+        for further information. Supplying contraints is optional.
+
+    Attributes
+    ----------
+    optimizer : bayes_opt.BayesianOptimization
+        The optimizer instance of this class.
+    """
+
+    verbose: InitVar[int] = 1
+    random_state: InitVar[int] = 1
+    allow_duplicate_points: InitVar[bool] = True
+    bayes_opt_kwds: InitVar[Optional[Mapping[str, Any]]] = None
+
+    def __post_init__(
+        self, verbose, random_state, allow_duplicate_points, bayes_opt_kwds
+    ):
+        """Create optimizer"""
+        if bayes_opt_kwds is None:
+            bayes_opt_kwds = {}
+
+        if self.input.bounds is None:
+            raise ValueError("Input.bounds is required for this optimizer")
+
+        self.optimizer = BayesianOptimization(
+            f=self._opt_func,
+            pbounds=self.input.bounds,
+            verbose=verbose,
+            random_state=random_state,
+            allow_duplicate_points=allow_duplicate_points,
+            **bayes_opt_kwds,
+        )
+
+    def _target_func(self, impact: Impact, data: pd.DataFrame) -> Number:
+        """Invert the cost function because BayesianOptimization maximizes the target"""
+        return 1 / self.input.cost_func(impact, data)
```

Comment:
This is not mathematically equivalent to minimizing the cost function. You should rather switch signs:
```suggestion
        return -self.input.cost_func(impact, data)
```
First of all, it is undefined if the cost function takes the value 0, but it also has very different slopes for large values so that convexity is lost.

If the user engineers the cost function in a way that it is convex and has good optimization properties, this transformation will break this.

Reply:
Absolutely right! I was afraid that something internal in the optimizer might break when using negative numbers, but it works just fine.